kotlinx-coroutines-core/common/src/Job.kt

```diff
@@ -361,7 +361,8 @@ internal fun Job.invokeOnCompletion(
     handler: InternalCompletionHandler
 ): DisposableHandle = when (this) {
     is JobSupport -> invokeOnCompletionInternal(onCancelling, invokeImmediately, handler)
-    else -> invokeOnCompletion(onCancelling, invokeImmediately, handler::invoke)
+    is NonCancellable -> NonDisposableHandle
+    else -> throw UnsupportedOperationException("third-party implementation of Job is not supported")
```

Comment:
This one might reasonably fail.
Could you please see if the pattern from https://github.com/Kotlin/kotlinx.coroutines/issues/2423 keeps working?

Reply:
It doesn't, the CI showed this. Reverted the change to `Job`.